metrics/metrics.go

```diff
@@ -37,44 +41,62 @@ var (
 
 	// Used in fallback checks, e.g. to avoid sending "counters" with 0 values
 	metricTypes map[MetricID]MetricType
+
+	// OTel metric instrumentation
+	meter    = otel.Meter("go.opentelemetry.io/ebpf-profiler")
+	counters = map[MetricID]metric.Int64Counter{}
+	gauges   = map[MetricID]metric.Int64Gauge{}
 )
 
 func init() {
-	defs, err := GetDefinitions()
-	if err != nil {
-		panic("extracting definitions from metrics.json")
-	}
-
+	defs := GetDefinitions()
 	metricTypes = make(map[MetricID]MetricType, len(defs))
 	for _, md := range defs {
+		if md.Obsolete {
+			continue
+		}
 		metricTypes[md.ID] = md.Type
+		switch typ := md.Type; typ {
+		case MetricTypeCounter:
+			counter, err := meter.Int64Counter(md.Name,
+				metric.WithDescription(md.Description),
+				metric.WithUnit(md.Unit))
+			if err != nil {
+				log.Errorf("Creating Int64Counter: %v", err)
+				continue
+			}
+			counters[md.ID] = counter
+		case MetricTypeGauge:
+			gauge, err := meter.Int64Gauge(md.Name,
+				metric.WithDescription(md.Description),
+				metric.WithUnit(md.Unit))
+			if err != nil {
+				log.Errorf("Creating Int64Gauge: %v", err)
+				continue
+			}
+			gauges[md.ID] = gauge
+		default:
+			panic(fmt.Sprintf("Unknown metric type: %v", typ))
+		}
 	}
 }
 
-// reporterImpl allows swapping out the global metrics reporter.
-//
-// nil is a valid value indicating that metrics should be voided.
-var reporterImpl reporter.MetricsReporter
-
-// SetReporter sets the reporter instance used to send out metrics.
-func SetReporter(r reporter.MetricsReporter) {
-	reporterImpl = r
-}
-
-// report converts and reports collected metrics via the reporter package.
+// report converts and reports collected metrics via OTel metrics
 func report() {
-	ids := make([]uint32, nMetrics)
-	values := make([]int64, nMetrics)
-
+	ctx := context.Background()
 	for i := 0; i < nMetrics; i++ {
-		ids[i] = uint32(metricsBuffer[i].ID)
-		values[i] = int64(metricsBuffer[i].Value)
-	}
-
-	if reporterImpl != nil {
-		reporterImpl.ReportMetrics(uint32(prevTimestamp), ids, values)
+		metric := metricsBuffer[i]
```

Comment:
Could we simplify this loop and just do a `for _, metric := range metricsBuffer {` instead?

Reply:
metricsBuffer is sized by worst-case ([IDMax](https://github.com/open-telemetry/opentelemetry-ebpf-profiler/blob/900e7e9f4bde37d88b7d7c961a53b07ec0fda035/metrics/metrics.go#L27)) and the number of metrics it actually contains is given by [nMetrics](https://github.com/open-telemetry/opentelemetry-ebpf-profiler/blob/900e7e9f4bde37d88b7d7c961a53b07ec0fda035/metrics/metrics.go#L34), so we can't iterate over the entire buffer.

Reply:
I roughly agree with Florian here.
`for _, metric := range metricsBuffer[0:nMetrics] {` would do it. Avoiding `i` here increases readability - otherwise when reading the code the first time, one tries to find if `i` is used in the loop body, which can be avoided.